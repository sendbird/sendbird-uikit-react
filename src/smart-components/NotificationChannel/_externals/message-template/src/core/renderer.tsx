import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
