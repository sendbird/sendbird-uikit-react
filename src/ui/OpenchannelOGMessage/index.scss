@import '../../styles/variables';

.sendbird-openchannel-og-message {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;

  &:hover {
    @include themed() {
      background-color: t(bg--1-4);
    }
    .sendbird-openchannel-og-message__top {
      .sendbird-openchannel-og-message__top__context-menu {
        display: block;
      }
    }
  }

  .sendbird-openchannel-og-message__top {
    box-sizing: border-box;
    position: relative;
    display: flex;
    flex-direction: row;

    .sendbird-openchannel-og-message__top__left {
      position: relative;
      display: flex;
      width: 52px;
      min-width: 52px;
  
      .sendbird-openchannel-og-message__top__left__avatar {
        position: absolute;
        display: flex;
        top: 8px;
        left: 24px;
      }
    }

    .sendbird-openchannel-og-message__top__right {
      position: relative;
      display: flex;
      box-sizing: border-box;
      flex-direction: column;
      margin: 8px 24px 8px 12px;
  
      .sendbird-openchannel-og-message__top__right__title {
        position: relative;
        display: flex;
        margin-bottom: 2px;
  
        .sendbird-openchannel-og-message__top__right__title__sender-name {
          position: relative;
          display: flex;
        }
        .sendbird-openchannel-og-message__top__right__title__sent-at {
          position: relative;
          display: flex;
          margin-left: 4px;
        }
      }

      .sendbird-openchannel-og-message__top__right__description {
        .sendbird-openchannel-og-message__top__right__description__message {
          position: relative;
          display: inline;
          white-space: pre-wrap;
          word-break: break-all;
          .sendbird-word .sendbird-word__url {
            @include themed() {
              color: t(primary-3);
            }
          }
        }
      }
    }

    .sendbird-openchannel-og-message__top__context-menu {
      position: absolute;
      right: 16px;
      display: none;
  
      .sendbird-openchannel-og-message__top__context-menu--icon {
        @include themed() {
          background-color: t(bg--2-5);
        }
        &:hover {
          @include themed() {
            background-color: t(primary--1-bg--7);
          }
        }
      }
    }
  }

  .sendbird-openchannel-og-message__bottom {
    position: relative;
    display: inline;
    margin-bottom: 8px;
    width: 100%;

    .sendbird-openchannel-og-message__bottom__og-tag {
      position: relative;
      display: inline-block;
      box-sizing: border-box;
      width: calc(100% - 64px);
      max-width: 360px;
      left: 64px;
      padding: 12px 12px 4px 12px;
      border-radius: 8px;
      @include themed() {
        border: solid 1px t(on-bg-4);
        background-color: t(bg-1);
      }
  
      .sendbird-openchannel-og-message__bottom__og-tag__url {
        position: relative;
        display: block;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        margin-bottom: 8px;
        max-width: 336px;
      }
      .sendbird-openchannel-og-message__bottom__og-tag__title {
        position: relative;
        display: inline-block;
        white-space: pre-wrap;
        overflow-wrap: normal;
        margin-bottom: 8px;
        max-width: 336px;
        &:hover {
          cursor: pointer;
          @include themed() {
            color: t(primary-3);
          }
        }
      }
      .sendbird-openchannel-og-message__bottom__og-tag__description {
        position: relative;
        display: block;
        white-space: pre-wrap;
        overflow: hidden;
        text-overflow: ellipsis;
        word-break: break-all;
        max-height: 32px;
        margin-bottom: 8px;
        width: 100%;
        max-width: 336px;
      }
      .sendbird-openchannel-og-message__bottom__og-tag__thumbnail {
        box-sizing: border-box;
        position: relative;
        display: block;
        margin-top: 4px;
        margin-bottom: 8px;
        width: 100%;
        &:hover {
          cursor: pointer;
        }
        .sendbird-openchannel-og-message__bottom__og-tag__thumbnail__image {
          border-radius: 8px;
          width: 100%;
          min-width: 200px;
          max-width: 336px;
        }
        .sendbird-openchannel-og-message__bottom__og-tag__thumbnail__image--placeholder {
          border-radius: 8px;
          width: 100%;
          min-width: 200px;
          max-width: 336px;
        }
      }
    }
  }

  .sendbird-openchannel-og-message__top__right__tail {
    position: relative;
    left: 64px;
    top: 2px;
  }
}

.sendbird-openchannel-og-message--word {
  display: inline;
  margin: 0px 2px;
  word-break: break-all;
}
