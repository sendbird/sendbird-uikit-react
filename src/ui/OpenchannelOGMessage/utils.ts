import Sendbird from 'sendbird';

export const URL_REG = /[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/;
export const createUrlTester = (regexp: RegExp) => (text: string): boolean => regexp.test(text);
export const checkOGIsEnalbed = (message: Sendbird.UserMessage): boolean => {
  const { ogMetaData } = message;
  if (!ogMetaData) {
    return false;
  }
  const { url } = ogMetaData;
  if (!url) {
    return false;
  }
  return true;
};
