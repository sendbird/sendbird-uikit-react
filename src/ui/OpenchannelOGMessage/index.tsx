import React, { useState, useContext, useMemo, useRef, useEffect } from 'react';
import { UserMessage } from '@sendbird/chat/message';
import format from 'date-fns/format';
import './index.scss';

import Avatar from '../Avatar';
import ContextMenu, { MenuItems, MenuItem } from '../ContextMenu';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import ImageRenderer from '../ImageRenderer';
import LinkLabel from '../LinkLabel';
import Label, { LabelTypography, LabelColors } from '../Label';
import Loader from '../Loader';
import UserProfile from '../UserProfile';
import { UserProfileContext } from '../../lib/UserProfileContext';

import uuidv4 from '../../utils/uuid';
import { copyToClipboard } from '../OpenchannelUserMessage/utils';
import { useLocalization } from '../../lib/LocalizationContext';
import { checkOGIsEnalbed, createUrlTester, URL_REG } from './utils';
import {
  checkIsPending,
  checkIsFailed,
  isFineCopy,
  isFineEdit,
  isFineResend,
  isFineDelete,
  showMenuTrigger,
} from '../../utils/openChannelUtils';
import { getSenderFromMessage } from '../../utils/openChannelUtils';

interface Props {
  message: UserMessage;
  isOperator?: boolean;
  className?: string | Array<string>;
  disabled?: boolean;
  showEdit(bool: boolean): void;
  showRemove(bool: boolean): void;
  resendMessage(message: UserMessage): void;
  chainTop?: boolean;
  chainBottom?: boolean;
  userId: string;
}

export default function OpenchannelOGMessage({
  message,
  isOperator,
  className,
  disabled,
  showEdit,
  showRemove,
  resendMessage,
  chainTop,
  userId,
}: Props): JSX.Element {
  if (!message || message.messageType !== 'user') {
    return null;
  }

  const status = message?.sendingStatus;
  const { ogMetaData } = message;
  const { defaultImage } = ogMetaData;

  const { stringSet, dateLocale } = useLocalization();
  const { disableUserProfile, renderUserProfile } = useContext(UserProfileContext);
  const [contextStyle, setContextStyle] = useState({});
  const messageComponentRef = useRef(null);
  const contextMenuRef = useRef(null);
  const avatarRef = useRef(null);

  const isUrl = createUrlTester(URL_REG);
  const isPending = checkIsPending(status);
  const isFailed = checkIsFailed(status);
  const sender = getSenderFromMessage(message);

  const MemoizedMessageText = useMemo(() => () => {
    const wordClassName = 'sendbird-openchannel-og-message--word';
    const splitMessage = message.message.split(' ');
    const matchedMessage = splitMessage
      .map((word) => (
        isUrl(word)
          ? (
            <LinkLabel
              key={uuidv4()}
              className={[
                wordClassName,
                'sendbird-openchannel-og-message--word--link'
              ]}
              src={word}
              type={LabelTypography.BODY_1}
              color={LabelColors.PRIMARY}
            >
              {word}
            </LinkLabel>
          )
          : (
            <Label
              key={uuidv4()}
              className={wordClassName}
              type={LabelTypography.BODY_1}
              color={LabelColors.ONBACKGROUND_1}
            >
              {word}
            </Label>
          )
      ));

    if (message.updatedAt > 0) {
      matchedMessage.push(
        <Label
          key={uuidv4()}
          className={wordClassName}
          type={LabelTypography.BODY_1}
          color={LabelColors.ONBACKGROUND_2}
        >
          {stringSet.MESSAGE_EDITED}
        </Label>,
      );
    }

    return matchedMessage;
  }, [message, message.updatedAt]);

  const openLink = () => {
    if (checkOGIsEnalbed(message)) {
      const { url } = ogMetaData;
      window.open(url);
    }
  };

  // place conxt menu top depending clientHeight of message component
  useEffect(() => {
    if (messageComponentRef?.current?.clientHeight > 36) {
      setContextStyle({ top: '8px ' });
    } else {
      setContextStyle({ top: '2px' });
    }
  }, [window.innerWidth]);

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-openchannel-og-message',
      ].join(' ')}
      ref={messageComponentRef}
    >
      <div
        className="sendbird-openchannel-og-message__top"
      >
        <div className="sendbird-openchannel-og-message__top__left">
          {
            !chainTop && (
              <ContextMenu
                menuTrigger={(toggleDropdown) => (
                  <Avatar
                    className="sendbird-openchannel-og-message__top__left__avatar"
                    src={sender.profileUrl || ''}
                    ref={avatarRef}
                    width="28px"
                    height="28px"
                    onClick={() => {
                      if (!disableUserProfile) {
                        toggleDropdown();
                      }
                    }}
                  />
                )}
                menuItems={(closeDropdown) => (
                  <MenuItems
                    parentRef={avatarRef}
                    parentContainRef={avatarRef}
                    closeDropdown={closeDropdown}
                    style={{ paddingTop: 0, paddingBottom: 0 }}
                  >
                    {
                      renderUserProfile
                        ? (
                          renderUserProfile({
                            user: sender,
                            close: closeDropdown,
                          })
                        )
                        : (
                          <UserProfile
                            user={sender}
                            onSuccess={closeDropdown}
                            disableMessaging
                          />
                        )
                    }
                  </MenuItems>
                )}
              />
            )
          }
        </div>
        <div className="sendbird-openchannel-og-message__top__right">
          {
            !chainTop && (
              <div className="sendbird-openchannel-og-message__top__right__title">
                <Label
                  className="sendbird-openchannel-og-message__top__right__title__sender-name"
                  type={LabelTypography.CAPTION_2}
                  color={isOperator ? LabelColors.SECONDARY_3 : LabelColors.ONBACKGROUND_2}
                >
                  {
                    sender && (
                      sender.friendName
                      || sender.nickname
                      || sender.userId
                    )
                  }
                </Label>
                <Label
                  className="sendbird-openchannel-og-message__top__right__title__sent-at"
                  type={LabelTypography.CAPTION_3}
                  color={LabelColors.ONBACKGROUND_3}
                >
                  {
                    message?.createdAt && (
                      format(message?.createdAt, 'p', {
                        locale: dateLocale,
                      })
                    )
                  }
                </Label>
              </div>
            )
          }
          <div className="sendbird-openchannel-og-message__top__right__description">
            <Label
              className="sendbird-openchannel-og-message__top__right__description__message"
              type={LabelTypography.BODY_1}
              color={LabelColors.ONBACKGROUND_1}
            >
              {MemoizedMessageText()}
            </Label>
          </div>
        </div>
        {
          <div
            className="sendbird-openchannel-og-message__top__context-menu"
            ref={contextMenuRef}
            style={contextStyle}
          >
            <ContextMenu
              menuTrigger={(toggleDropdown) => (
                showMenuTrigger({ message: message, userId: userId, status: status }) && (
                  <IconButton
                    className="sendbird-openchannel-og-message__top__context-menu--icon"
                    width="32px"
                    height="32px"
                    onClick={() => {
                      toggleDropdown();
                    }}
                  >
                    <Icon
                      type={IconTypes.MORE}
                      fillColor={IconColors.CONTENT_INVERSE}
                      width="24px"
                      height="24px"
                    />
                  </IconButton>
                )
              )}
              menuItems={(closeDropdown) => (
                <MenuItems
                  parentRef={contextMenuRef}
                  parentContainRef={contextMenuRef}
                  closeDropdown={closeDropdown}
                  openLeft
                >
                  {
                    isFineCopy({ message, userId, status }) && (
                      <MenuItem
                        className="sendbird-openchannel-og-message__top__context-menu__copy"
                        onClick={() => {
                          copyToClipboard(message.message);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__COPY}
                      </MenuItem>
                    )
                  }
                  {
                    isFineEdit({ message, userId, status }) && (
                      <MenuItem
                        className="sendbird-openchannel-og-message__top__context-menu__edit"
                        onClick={() => {
                          if (disabled) {
                            return;
                          }
                          showEdit(true);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__EDIT}
                      </MenuItem>
                    )
                  }
                  {
                    isFineResend({ message, userId, status }) && (
                      <MenuItem
                        className="sendbird-openchannel-og-message__top__context-menu__resend"
                        onClick={() => {
                          resendMessage(message);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__RESEND}
                      </MenuItem>
                    )
                  }
                  {
                    isFineDelete({ message, userId, status }) && (
                      <MenuItem
                        className="sendbird-openchannel-og-message__top__context-menu__delete"
                        onClick={() => {
                          if (disabled) {
                            return;
                          }
                          showRemove(true);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__DELETE}
                      </MenuItem>
                    )
                  }
                </MenuItems>
              )}
            />
          </div>
        }
      </div>
      <div className="sendbird-openchannel-og-message__bottom">
        <div className="sendbird-openchannel-og-message__bottom__og-tag">
          {
            ogMetaData.url && (
              <Label
                className="sendbird-openchannel-og-message__bottom__og-tag__url"
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_2}
              >
                {ogMetaData.url}
              </Label>
            )
          }
          {
            ogMetaData.title && (
              <LinkLabel
                className="sendbird-openchannel-og-message__bottom__og-tag__title"
                src={ogMetaData.url}
                type={LabelTypography.SUBTITLE_2}
                color={LabelColors.PRIMARY}
              >
                {
                  ogMetaData.title
                }
              </LinkLabel>
            )
          }
          {
            ogMetaData.description && (
              <Label
                className="sendbird-openchannel-og-message__bottom__og-tag__description"
                type={LabelTypography.BODY_2}
                color={LabelColors.ONBACKGROUND_1}
              >
                {ogMetaData.description}
              </Label>
            )
          }
          {
            ogMetaData.url && (
              <div
                className="sendbird-openchannel-og-message__bottom__og-tag__thumbnail"
                role="button"
                onClick={openLink}
                onKeyDown={openLink}
                tabIndex={0}
              >
                {
                  defaultImage && (
                    <ImageRenderer
                      className="sendbird-openchannel-og-message__bottom__og-tag__thumbnail__image"
                      url={defaultImage.url || ''}
                      alt={defaultImage.alt || ''}
                      height="189px"
                      defaultComponent={(
                        <div className="sendbird-openchannel-og-message__bottom__og-tag__thumbnail__image--placeholder">
                          <Icon
                            type={IconTypes.THUMBNAIL_NONE}
                            width="56px"
                            height="56px"
                          />
                        </div>
                      )}
                    />
                  )
                }
              </div>
            )
          }
        </div>
        {
          (isPending || isFailed) && (
            <div className="sendbird-openchannel-og-message__top__right__tail">
              {
                isPending && (
                  <Loader
                    width="16px"
                    height="16px"
                  >
                    <Icon
                      className="sendbird-openchannel-og-message__top__right__tail__pending"
                      type={IconTypes.SPINNER}
                      fillColor={IconColors.PRIMARY}
                      width="16px"
                      height="16px"
                    />
                  </Loader>
                )
              }
              {
                isFailed && (
                  <Icon
                    className="sendbird-openchannel-og-message__top__right__tail__failed"
                    type={IconTypes.ERROR}
                    fillColor={IconColors.ERROR}
                    width="16px"
                    height="16px"
                  />
                )
              }
            </div>
          )
        }
      </div>
    </div>
  );
}
