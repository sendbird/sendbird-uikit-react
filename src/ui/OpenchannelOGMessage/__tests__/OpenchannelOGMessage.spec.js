import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import OpenchannelOGMessage from "../index";

// mock date-fns to avoid problems from snapshot timestamping
// between testing in different locations
// ideally we want to mock date-fns globally - needs more research
jest.mock('date-fns/format', () => () => ('mock-date'));

const userId = 'hh-1234';
const getMockMessage = (callback) => {
  const message = {
    messageType: 'user',
    message: 'I am the Message',
    createdAt: 1111,
    updatedAt: 0,
    ogMetaData: {
      url: 'https://sendbird.com/',
      title: 'This is the TITLE',
      description: 'I am description I am who has much string in this og meta data',
      defaultImage: {
        url: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
        alt: 'test',
      },
    },
    sender: {
      profileUrl: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
      friendName: 'Hoon Baek',
      nickname: 'Honn',
      userId,
    },
    isResendable: () => false,
  };
  if (callback) {
    return callback(message);
  }
  return message;
};

describe('OpenchannelOGMessage', () => {
  it('should have default elements', function() {
    const component = mount(
      <OpenchannelOGMessage
        message={getMockMessage()}
        userId={userId}
        status="succeeded"
      />
    );

    expect(
      component.find('.sendbird-openchannel-og-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__description__message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__url').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__title').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__description').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__thumbnail').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should not have elements by chainTop', function() {
    const component = mount(
      <OpenchannelOGMessage
        message={getMockMessage()}
        userId={userId}
        status="succeeded"
        chainTop
      />
    );

    expect(
      component.find('.sendbird-openchannel-og-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left__avatar').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sender-name').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sent-at').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__description__message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__url').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__title').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__description').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__thumbnail').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should not have og elements when ogMetaData does not exist', function() {
    const component = mount(
      <OpenchannelOGMessage
        message={getMockMessage((message) => ({ ...message, ogMetaData: {} }))}
        status="succeeded"
        userId="hh-1234"
      />
    );

    expect(
      component.find('.sendbird-openchannel-og-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__description__message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__url').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__title').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__description').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__bottom__og-tag__thumbnail').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render pending icon if status is pending', function() {
    const component = mount(
      <OpenchannelOGMessage
        message={getMockMessage((message) => ({ ...message, sendingStatus: 'pending' }))}
      />
    );
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__pending').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render failed icon if status is failed', function() {
    const component = mount(
      <OpenchannelOGMessage
        message={getMockMessage((message) => ({ ...message, sendingStatus: 'failed' }))}
        status="failed"
      />
    );
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-og-message__top__right__tail__failed').exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the OpenchannelOGMessage DOM', function() {
    const component = renderer.create(
      <OpenchannelOGMessage
        message={{
          messageType: 'user',
          message: 'I am the Message',
          createdAt: 1111,
          updatedAt: 0,
          ogMetaData: {
            url: 'https://sendbird.com/',
            title: 'This is the TITLE',
            description: 'I am description I am who has much string in this og meta data',
            defaultImage: {
              url: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
              alt: 'test',
            },
          },
          sender: {
            profileUrl: 'https://static.sendbird.com/sample/profiles/profile_12_512px.png',
            friendName: 'Hoon Baek',
            nickname: 'Honn',
            userId: 'hh-1234',
          },
          isResendable: () => false,
        }}
        status="succeeded"
        userId="hh-1234"
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
