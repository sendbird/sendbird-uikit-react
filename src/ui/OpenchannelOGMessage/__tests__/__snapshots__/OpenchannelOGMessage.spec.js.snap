// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/OpenchannelOGMessage should do a snapshot test of the OpenchannelOGMessage DOM 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-openchannel-og-message"
  >
    <div
      class="sendbird-openchannel-og-message__top"
    >
      <div
        class="sendbird-openchannel-og-message__top__left"
      >
        <div
          class="sendbird-context-menu"
          style="display: inline;"
        >
          <div
            class="sendbird-openchannel-og-message__top__left__avatar sendbird-avatar"
            role="button"
            style="height: 28px; width: 28px; z-index: 0; left: 0px;"
            tabindex="0"
          >
            <div
              class="sendbird-avatar-img sendbird-image-renderer"
              style="width: 100%; min-width: min(400px, 28px); max-width: 400px;"
            >
              <div
                class="sendbird-image-renderer__image"
                style="width: 100%; min-width: min(400px, 28px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_12_512px.png);"
              />
              <img
                alt=""
                class="sendbird-image-renderer__hidden-image-loader"
                src="https://static.sendbird.com/sample/profiles/profile_12_512px.png"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sendbird-openchannel-og-message__top__right"
      >
        <div
          class="sendbird-openchannel-og-message__top__right__title"
        >
          <span
            class="sendbird-openchannel-og-message__top__right__title__sender-name sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-2"
          >
            Hoon Baek
          </span>
          <span
            class="sendbird-openchannel-og-message__top__right__title__sent-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-3"
          >
            mock-date
          </span>
        </div>
        <div
          class="sendbird-openchannel-og-message__top__right__description"
        >
          <span
            class="sendbird-openchannel-og-message__top__right__description__message sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
          >
            I am the Message
          </span>
        </div>
      </div>
      <div
        class="sendbird-openchannel-og-message__top__context-menu"
        style="top: 2px;"
      >
        <div
          class="sendbird-context-menu"
          style="display: inline;"
        >
          <button
            class="sendbird-openchannel-og-message__top__context-menu--icon sendbird-iconbutton "
            style="height: 32px; width: 32px;"
            type="button"
          >
            <span
              class="sendbird-iconbutton__inner"
            >
              <div
                class=" sendbird-icon sendbird-icon-more sendbird-icon-color--content-inverse"
                role="button"
                style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
                tabindex="0"
              >
                <test-file-stub />
              </div>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-og-message__bottom"
    >
      <div
        class="sendbird-openchannel-og-message__bottom__og-tag"
      >
        <span
          class="sendbird-openchannel-og-message__bottom__og-tag__url sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
        >
          https://sendbird.com/
        </span>
        <a
          class="sendbird-openchannel-og-message__bottom__og-tag__title sendbird-link-label sendbird-label--color-primary"
          href="https://sendbird.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="sendbird-link-label__label sendbird-label sendbird-label--subtitle-2 sendbird-label--color-primary"
          >
            This is the TITLE
          </span>
        </a>
        <span
          class="sendbird-openchannel-og-message__bottom__og-tag__description sendbird-label sendbird-label--body-2 sendbird-label--color-onbackground-1"
        >
          I am description I am who has much string in this og meta data
        </span>
        <div
          class="sendbird-openchannel-og-message__bottom__og-tag__thumbnail"
          role="button"
          tabindex="0"
        >
          <div
            class="sendbird-openchannel-og-message__bottom__og-tag__thumbnail__image sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, 400px); max-width: 400px;"
          >
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, 400px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_12_512px.png);"
            />
            <img
              alt="test"
              class="sendbird-image-renderer__hidden-image-loader"
              src="https://static.sendbird.com/sample/profiles/profile_12_512px.png"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
